src/wasm/wasm-type.cpp

```diff
@@ -3306,13 +3659,39 @@ std::optional<TypeBuilder::Error> canonicalizeIsorecursive(
     return {*error};
   }
 
-  // Move the recursion groups into the global store. TODO: after proper
-  // isorecursive canonicalization, some groups may no longer be used, so they
-  // will need to be filtered out.
-  std::lock_guard<std::mutex> lock(recGroupsMutex);
-  for (auto& info : recGroupInfos) {
-    recGroups.emplace_back(std::move(info));
+  // Now that we know everything is valid, start canonicalizing recursion
+  // groups. Before canonicalizing each group, update all the use sites with it
+  // to make sure it only refers to other canonical groups or to itself. To
+  // canonicalize the group, try to insert it into the global store. If that
+  // fails, we already have an isorecursively equivalent group, so update the
+  // replacements accordingly.
+  CanonicalizationState::ReplacementMap replacements;
+  {
+    std::lock_guard<std::mutex> lock(globalRecGroupStore.mutex);
+    groupStart = 0;
+    for (auto group : groups) {
+      size_t size = group.size();
+      for (size_t i = 0; i < size; ++i) {
+        state.updateUses(replacements, state.newInfos[groupStart + i]);
+      }
+      groupStart += size;
+      RecGroup canonical(0);
+      if (auto it = groupInfoMap.find(group); it != groupInfoMap.end()) {
+        // Move ownership to the global store if this group is canonical.
```

Comment:
I don't get this. `groupInfoMap` is not a place where presence indicates being canonical or not, does it? The entire canonicalization aspect is handled a few lines down where you do `if (group != canonical)`, I think. Then why does this line mention "canonical"? I am also puzzled by the word. Does "is canonical" mean "is the canonical one" or "is canonicalized to something else"?

Reply:
> Does "is canonical" mean "is the canonical one" or "is canonicalized to something else"?

The former.

`globalRecGroupStore.insert(std::unique_ptr<RecGroupInfo>&&)`, called on the line after this comment, moves the info into the store if and only if it is the info for the first rec group of this structure to be inserted, i.e. if the rec group becomes the canonical rec group for its structure. I'll rename the method `maybeInsert` to make it clearer that it only conditionally takes ownership of its argument.

Reply:
But actually this is already consistent with how the `insert` methods of `globalHeapTypeStore` and `globalTypeStore` already work, so maybe it would be better to keep the naming consistent (or to change it everywhere in a follow up).

Reply:
I updated the comment to hopefully make it clearer.

Reply:
I see, existing naming seems ok then.

I am still having trouble with the logic, though. `auto it = groupInfoMap.find(group); it != groupInfoMap.end()` returns true if we find the item in that map. If we find it, doesn't that mean it is *not* canonical? But the move is in that arm of the if.

Reply:
This condition is about whether or not there exists a `RecGroupInfo` for this group, in which case it would live in the `groupInfoMap`. The alternative is that this is a singleton group so there is no `RecGroupInfo`. In either case, the group may or may not end up being the canonical group for its structure.

Reply:
I see, thanks. I missed that this was about the singleton issue...  code is quite complex so I'd suggest a comment about that as well, as the current ones just speak about canonicalization here.

Reply:
Added a bigger comment about info ownership to replace this comment.

Reply:
"If it is canonical" - based on your last comment, shouldn't it be "is not a singleton"?